Added CountingVariantFilter

magicDGS opened the pull request:
Following the same implementation as `CountingReadFilter`, this is a way of counting the `VariantContext` that were filtered. This is useful for report the status of `VariantFilter`s.

There are not tests yet because I would like to have the ok on this before spending time on implement them. I think that the community could benefit from using this, and even some tools from the GATK itself.

droazen commented:
@cmnbroad could you review this one, since you implemented the current read filtering framework?

cmnbroad commented:
@magicDGS I like your suggestion of factoring out a CountingFilter class that can be reused for both filter types. Ideally, I think we should first get https://github.com/broadinstitute/gatk/pull/2218/files in first (it has some minor changes to CountingReadFilter). Hopefully that will be soon. In the meantime, if you're so inclined, you might try doing the CountingFilter refactoring as part of this PR (which still needs tests), since you'll have two clients for it. That will definitely require some careful refactoring of the existing classes in order to retain the current CountingFilter behavior).

magicDGS commented:
Thanks for the feedback @cmnbroad! I could wait til #2218 is accepted to continue working and check that nothing is broked...

cmnbroad commented:
@magicDGS #2218 is merged now.

cmnbroad commented:
@magicDGS BTW, its probably possible to factor out a generic CountingFilter, but it may not be worth the cost in complexity required. I think the type constraints are (for X in {ReadFilter, VariantFilter}):

- an Xfilter is-a predicate
- a CountingXFilter is-an XFilter
- any filter resulting from AND/OR’ing a CountingXFilter also is-a CountingXFilter
- CountingXFilter reports counts at each granular level

If you have a solution that satisfies these constraints an isn't overly complicated, that would be great! But if not, a separate CountingVariantFilter is also fine.

cmnbroad commented:
Hi @magicDGS. I went down this same path (composition of CountingFilter) when I originally implemented CountingReadFilter, but I abandoned it for the model we currently have. I think the current model is much simpler in a number of ways. This is an interesting problem, but I would say lets just implement a straight CountingVariantFilter/tests and not try to do the common implementation. 

BTW, when looking at this PR I noticed two bugs in the existing code that we should make sure not to propagate to the Variant one (feel free to fix/test these as part of this PR):

- CountingReadFilter.resetFilterCount only resets the root filter count; it needs an override in BinOp to propagate the reset call to the lhs/rhs operands.
- there is a bug in the getSummary tests/code; you can see the fix [here](https://github.com/broadinstitute/gatk/commit/9ef1458271834aed9b64a5d66f94df33f025eafb)

Thanks!

magicDGS commented:
@cmnbroad, I opened a PR to fix `CountingReadFilter` and here I push two commits: one removing the complicated implementation and comming back to an independent `CountingVariantFilter`and the second with the tests. Can you review?

magicDGS commented:
That should be now ready to review, @cmnbroad! I rebased to the latest master and now this should pass the tests.

cmnbroad commented on src/test/java/org/broadinstitute/hellbender/engine/filters/CountingVariantUnitTest.java:
```diff
@@ -0,0 +1,308 @@
+
+public final class CountingVariantUnitTest {
+
+    // Mirrors CountingVariantFilterUnitTest
```
This file *is* the counting variant unit test - I would just remove this comment since it was relevant in the counting read filter tests, but its not here.)

magicDGS replied:
Copy-paste issue coming from CountingReadFilterUnitTest. Removed!

magicDGS commented on src/test/java/org/broadinstitute/hellbender/engine/filters/CountingVariantUnitTest.java:
```diff
@@ -0,0 +1,308 @@
+
+import java.util.*;
+
+public final class CountingVariantUnitTest {
```
Sure, I think that this was an error caused by some refactoring/rebase... I can't actually remember....

magicDGS commented on src/test/java/org/broadinstitute/hellbender/engine/filters/CountingVariantUnitTest.java:
```diff
@@ -0,0 +1,308 @@
+        Assert.assertTrue(vf.getClass() == CountingVariantFilter.CountingAndVariantFilter.class);
+        CountingVariantFilter.CountingAndVariantFilter andFilter = (CountingVariantFilter.CountingAndVariantFilter) vf;
+
+        // lhs is a Counting and filter; rhs is a counting filter that delegates to GOOD_CIGAR
```
Done.

magicDGS commented on src/test/java/org/broadinstitute/hellbender/engine/filters/CountingVariantUnitTest.java:
```diff
@@ -0,0 +1,308 @@
+        andFilter = (CountingVariantFilter.CountingAndVariantFilter) andFilter.lhs;
+
+        // lhs is a Counting filter that delegates to MAPPING_QUALITY_AVAILABLE; rhs is a
+        // counting filter that delegates to MAPPED
```
Done.

magicDGS commented:
Addressed comments - back to you @cmnbroad!

I rebased into the latest master, but not squashed to keep track of the changes. I think that it could be squashed at the same time as merging, couldn't it? let me know if you also need me to squash...
